lib/puppet/configurer.rb

```diff
@@ -328,6 +353,35 @@ def run_internal(options)
   end
   private :run_internal
 
+  def find_functional_server()
+    configured_environment = Puppet[:environment] if Puppet.settings.set_by_config?(:environment)
+
+    node = nil
+    selected_server = Puppet.settings[:server_list].find do |server|
+      # Puppet.override doesn't return the result of its block, so we
+      # need to handle this manually
+      found = false
+      server[1] ||= Puppet[:masterport]
```

Comment:
I just moved this "missing port" logic from the `run` method, since we need a port in order to contact the master when checking if it exists. This fixes a bug @Magisus found when working on the reporting aspect of this.